### PR TITLE
[REVIEW] Add File IO to cuIO benchmarks

### LINKED ISSUE #6390: [FEA] cuIO benchmarks do not use file IO
The C++ cuIO benchmarks do not perform file IO, and so the performance of this part of the implementation is not measured.

Proposed solution: by default, read/write files using a temp directory, and include an option to use other source/sink types (e.g. when SOL performance is measured).

### PR DESCRIPTION
**Pull request by vuule**
Closes #6390 

Adds an argument to all cuIO benchmarks that selects the type of IO to be used (host buffer/file/void). 
By default, all applicable IO types are used. Users can select a specific type using the `--benchmark-filter` command line argument with value 'buffer' or 'file' or 'void' (writers only).

Adds a utility class for creating coupled `source_info` and `sink_info` objects.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Review by kkraus14 — APPROVED** [review]
cmake lgtm

**Comment by vuule** [comment]
rerun tests

File: cpp/include/cudf_test/base_fixture.hpp [hunk 1]
```diff
@@ -20,6 +20,7 @@
 #include <cudf/utilities/traits.hpp>
 #include <cudf_test/cudf_gtest.hpp>
 #include <cudf_test/cxxopts.hpp>
+#include <tests/utilities/file_utilities.hpp>
```
Comment (devavret) [inline cpp/include/cudf_test/base_fixture.hpp:23]: Why is this file not in cudf_test

File: cpp/tests/utilities/file_utilities.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    if (const char *env_p = std::getenv("WORKSPACE")) dir_template = env_p;
+    dir_template += "/" + base_name + ".XXXXXX";
+    auto const tmpdirptr = mkdtemp(const_cast<char *>(dir_template.data()));
+    if (tmpdirptr == nullptr) CUDF_FAIL("Temporary directory creation failure: " + dir_template);
+    _path = dir_template + "/";
+  }
+
+  static int rm_files(const char *pathname, const struct stat *sbuf, int type, struct FTW *ftwb)
+  {
+    return remove(pathname);
```
Comment (davidwendt) [inline cpp/tests/utilities/file_utilities.hpp:42]:
```suggestion
    return std::remove(pathname);
```

**Review by davidwendt — APPROVED** [review]
Looks good to me.

File: cpp/include/cudf_test/base_fixture.hpp (hunk 1 above)
Comment (vuule) [reply cpp/include/cudf_test/base_fixture.hpp:23]: Messed the pattern. I'll move it.

File: cpp/benchmarks/io/cuio_benchmark_common.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <benchmarks/io/cuio_benchmark_common.hpp>
+
+#include <cstdio>
+#include <iostream>
```
Comment (karthikeyann) [inline cpp/benchmarks/io/cuio_benchmark_common.cpp:20]: clean up includes once.
some includes may not be needed anymore.

File: cpp/include/cudf_test/file_utilities.hpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    if (const char *env_p = std::getenv("WORKSPACE")) dir_template = env_p;
+    dir_template += "/" + base_name + ".XXXXXX";
+    auto const tmpdirptr = mkdtemp(const_cast<char *>(dir_template.data()));
+    if (tmpdirptr == nullptr) CUDF_FAIL("Temporary directory creation failure: " + dir_template);
+    _path = dir_template + "/";
+  }
+
+  static int rm_files(const char *pathname, const struct stat *sbuf, int type, struct FTW *ftwb)
+  {
+    return std::remove(pathname);
```
Comment (karthikeyann) [inline cpp/include/cudf_test/file_utilities.hpp:42]: `#include <cstdio>`
required include is included in another file. check all std include once.

File: cpp/include/cudf_test/file_utilities.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#include <cudf/utilities/error.hpp>
+
+class temp_directory {
+  std::string _path;
+
+ public:
+  temp_directory(const std::string &base_name)
+  {
+    std::string dir_template("/tmp");
+    if (const char *env_p = std::getenv("WORKSPACE")) dir_template = env_p;
```
Comment (karthikeyann) [inline cpp/include/cudf_test/file_utilities.hpp:33]: `#include <cstdlib>` for `std::getenv`

**Review by karthikeyann — CHANGES_REQUESTED** [review]
includes cleanup only.

File: cpp/benchmarks/io/cuio_benchmark_common.cpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+#include <string>
+
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/benchmarks/io/cuio_benchmark_common.cpp:23]: needed for `close`.

**Review by karthikeyann — APPROVED** [review]
LGTM

**Review by mike-wendt — APPROVED** [review]
Approved changes for our `CODEOWNERS` files; will rely on team to verify other changes

**Pull request merged**